test/xds/xds_client_custom_lb_test.go

```diff
@@ -0,0 +1,243 @@
+						TypedConfig: testutils.MarshalAny(m),
+					},
+				},
+			},
+		},
+	}
+}
+
+// clusterWithLBConfiguration returns a cluster resource with the proto message
+// passed Marshaled to an any and specified through the load_balancing_policy
+// field.
+func clusterWithLBConfiguration(clusterName, edsServiceName string, secLevel e2e.SecurityLevel, m proto.Message) *v3clusterpb.Cluster {
+	cluster := e2e.DefaultCluster(clusterName, edsServiceName, secLevel)
+	cluster.LoadBalancingPolicy = &v3clusterpb.LoadBalancingPolicy{
+		Policies: []*v3clusterpb.LoadBalancingPolicy_Policy{
+			{
+				TypedExtensionConfig: &v3corepb.TypedExtensionConfig{
+					Name:        "noop name",
+					TypedConfig: testutils.MarshalAny(m),
+				},
+			},
+		},
+	}
+	return cluster
+}
+
+// clientResourcesNewFieldSpecifiedAndPortsInMultipleLocalities returns default
+// xDS resources with two localities, of weights 1 and 2 respectively. It must
+// be passed 5 ports, and the first two ports will be put in the first locality,
+// and the last three will be put in the second locality. It also configures the
+// proto message passed in as the Locality + Endpoint picking policy in CDS.
+func clientResourcesNewFieldSpecifiedAndPortsInMultipleLocalities2(params e2e.ResourceParams, ports []uint32, m proto.Message) e2e.UpdateOptions {
+	routeConfigName := "route-" + params.DialTarget
+	clusterName := "cluster-" + params.DialTarget
+	endpointsName := "endpoints-" + params.DialTarget
+	return e2e.UpdateOptions{
+		NodeID:    params.NodeID,
+		Listeners: []*v3listenerpb.Listener{e2e.DefaultClientListener(params.DialTarget, routeConfigName)},
+		Routes:    []*v3routepb.RouteConfiguration{e2e.DefaultRouteConfig(routeConfigName, params.DialTarget, clusterName)},
+		Clusters:  []*v3clusterpb.Cluster{clusterWithLBConfiguration(clusterName, endpointsName, params.SecLevel, m)},
+		Endpoints: []*v3endpointpb.ClusterLoadAssignment{e2e.EndpointResourceWithOptionsMultipleLocalities(e2e.EndpointOptions{
+			ClusterName: endpointsName,
+			Host:        params.Host,
+			PortsInLocalities: [][]uint32{
+				{ports[0], ports[1]},
+				{ports[2], ports[3], ports[4]},
+			},
+			LocalityWeights: []uint32{
+				1,
+				2,
+			},
+		})},
+	}
+}
+
+// TestWRRLocality tests RPC distribution across a scenario with 5 backends,
+// with 2 backends in a locality with weight 1, and 3 backends in a second
+// locality with weight 2. Through xDS, the test configures a
+// wrr_locality_balancer with either a round robin or custom (specifying pick
+// first) child load balancing policy, and asserts the correct distribution
+// based on the locality weights and the endpoint picking policy specified.
+func (s) TestWrrLocality(t *testing.T) {
+	oldCustomLBSupport := envconfig.XDSCustomLBPolicy
+	envconfig.XDSCustomLBPolicy = true
+	defer func() {
+		envconfig.XDSCustomLBPolicy = oldCustomLBSupport
+	}()
+
+	managementServer, nodeID, _, r, cleanup := e2e.SetupManagementServer(t, e2e.ManagementServerOptions{})
```

Comment:
I would place the creation of the management server and the backends into `t.Run` as well (to guarantee maximum separation between the tests). Now the sub tests share the same management server and backends which could lead to some interesting flakes which will not be fun to debug.

Reply:
I have to spin up backends before since backends are dynamically chosen the address and I use that as expectation. I made the management server per test though. The backends simply exist and respond to requests, no state persisted, so I think that's fine to keep for each iteration.